apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+     calls to ReactDOM.render() are not)
+   - does not interact with remote, stateful services (ajax calls to read-only
+     services could be OK)
+   - does not use or provide shared application state (at least not in a direct way)
+
+   Some examples of "capsule libraries" in our code base would be
+   `src/MusicController.js`, `src/dom.js`, `src/Sounds.js`.
+
+
+## Capsule Library ##
+
+### Capsule Library Pattern ###
+
+In the case of a relatively simple capsule library that consists of only one
+file, the pattern is pretty simple:
+
+- Export functions and classes that other libraries are allowed to use. Do not
+  change the interfaces of these functions/classes in a way that is not
+  backwards compatible without updating all the callsites.
+- If the library works with internal state, encapsulate that state into a class
+  that must be instantiated by the calling code.
+- If the library works with internal state, but only supports a singleton
+  instance, only export the singleton instance and not the backing class.
+
+For more complicated capsule libraries whose code is spread across multiple
+files the pattern is only a bit more involved:
+
+- Put all the library files into a directory, and expose a single index.js file
+  which exports the functions/classes that other libraries are allowed to
+  use. Calling code imports from the index.js file only. This makes dependency
+  graphs easier to read.
+
+### Capsule Library Rules ###
+
+The rules for capsule libraries are also straightforward:
+
+- Do not depend on feature libraries
+- Do not read from or write to the global window object (accessing standard browser
+  APIs is OK)
+- Do not reference DOM nodes by ID or css selectors that are not passed in to
+  the library explicitly.
+
+
+## Feature Library ##
+
+Feature libraries require more discipline to avoid coupling and spaghetti code,
+since they must inherently interact with some form of global state at some
+point. But if we are careful and consistently apply a sensible pattern for
+writing feature libraries, we can keep things clean and manageable.
+
+### Feature Library Pattern ###
+
+The primary goal of the feature library pattern is to make it really clear where
+and how the library interacts with shared state or remote services.
+
+#### Exposing/Providing Shared State ####
+
+Shared application state should be managed by redux. Since there can only be one
+redux store on a page, while there can be multiple features on a page, feature
+libraries are not allowed to create their own redux store. However, they can
+have their own actions and reducers which operate on a well defined portion of
+the global redux state tree. https://github.com/erikras/ducks-modular-redux
+describes one pattern for doing this, which we will riff on.
+
+Feature libraries that expose shared redux state should have the following
+directory structure:
+
+```
+src/
+  <feature-name>/
+    index.js
```

Comment:
in that case we could replace `index.js` with whatever the name of the feature is, e.g. `src/dataBrowser/dataBrowser.js`.